src/main/java/com/gooddata/warehouse/WarehouseService.java

```diff
@@ -492,56 +494,103 @@ public WarehouseS3Credentials getWarehouseS3Credentials(final Warehouse warehous
+            throw new GoodDataException(format("Empty response when trying to %s S3 credentials via API", httpMethod.name()));
         }
 
-        return new PollResult<>(this,
-                new AbstractPollHandler<WarehouseTask, WarehouseS3Credentials>
-                        (task.getPollUri(), WarehouseTask.class, WarehouseS3Credentials.class) {
-
-                    @Override
-                    public boolean isFinished(ClientHttpResponse response) throws IOException {
-                        return HttpStatus.CREATED.equals(response.getStatusCode());
-                    }
-
-                    @Override
-                    public void handlePollResult(WarehouseTask pollResult) {
-                        try {
-                            final WarehouseS3Credentials created = restTemplate
-                                    .getForObject(pollResult.getWarehouseS3CredentialsUri(),
-                                            WarehouseS3Credentials.class);
-                            setResult(created);
-                        } catch (GoodDataException | RestClientException e) {
-                            throw new GoodDataException(
-                                    "S3 credentials added to warehouse, but can't get them back, uri: "
-                                            + pollResult.getWarehouseUserUri(), e);
-                        }
-                    }
-
-                    @Override
-                    public void handlePollException(final GoodDataRestException e) {
-                        throw new GoodDataException("Unable to add S3 credentials to warehouse, uri: "
-                                + warehouse.getUri(), e);
-                    }
-                });
+        return taskHttpEntity.getBody();
+    }
+
+    private AbstractPollHandler<WarehouseTask, WarehouseS3Credentials> getS3PollHandler(final WarehouseS3Credentials s3Credentials,
+                                                                                        final WarehouseTask task,
+                                                                                        final String action) {
+        return getS3PollHandler(s3Credentials, task, WarehouseS3Credentials.class, action);
+    }
+
+    private <R> AbstractPollHandler<WarehouseTask, R> getS3PollHandler(final WarehouseS3Credentials s3Credentials,
+                                                                       final WarehouseTask task,
+                                                                       final Class<R> resultClass,
+                                                                       final String action) {
+        return new AbstractPollHandler<WarehouseTask, R>(task.getPollUri(), WarehouseTask.class, resultClass) {
+
+            @Override
+            public boolean isFinished(ClientHttpResponse response) throws IOException {
+                return HttpStatus.CREATED.equals(response.getStatusCode());
```

Comment:
changed status of finished async `DELETE` and `PUT` tasks to `200 OK` in gooddata/gdc-datawarehouse#999 and in this PR also.